File: kotlin/src/main/kotlin/Schemas/DBModels/Contract.kt
package dev.threadify.Schemas.DBModels

import dev.threadify.DAOs.ContractDAO
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.JsonObject
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.dao.Entity
import org.jetbrains.exposed.dao.EntityClass
import org.jetbrains.exposed.dao.id.UUIDTable
import org.jetbrains.exposed.sql.javatime.timestamp
import java.time.Instant
import java.util.UUID

/**
 * Exposed Table definition for contracts
 */
object Contracts : UUIDTable("contracts") {
    val name = varchar("name", 255)
    val description = text("description")
    val contentHash = varchar("content_hash", 64).nullable()
    val latestVersion = integer("latest_version").default(1)
    val ownerId = varchar("owner_id", 255)
    val isPublic = bool("is_public").default(false)
    val isDeleted = bool("is_deleted").default(false)
    val createdAt = timestamp("created_at").default(Instant.now())
    val updatedAt = timestamp("updated_at").default(Instant.now())

    init {
      // Unique constraint: one contract name per owner id
      uniqueIndex("unique_contract_name", name, ownerId)
    }
}

/**
 * Exposed DAO Entity for Contract
 */
class Contract(id: EntityID<UUID>) : Entity<UUID>(id) {
    companion object : EntityClass<UUID, Contract>(Contracts)
    
    var name by Contracts.name
    var description by Contracts.description
    var contentHash by Contracts.contentHash
    var latestVersion by Contracts.latestVersion
    var ownerId by Contracts.ownerId
    var isPublic by Contracts.isPublic
    var isDeleted by Contracts.isDeleted
    var createdAt by Contracts.createdAt
    var updatedAt by Contracts.updatedAt
    
    // Relationship: one contract has many versions
    val versions by ContractVersion referrersOn ContractVersions.contractId
}

fun Contract.toDAO(includes: List<String> = emptyList()): ContractDAO {
 // Eagerly access all properties within the transaction
 val idValue = id.value.toString()
 val nameValue = name
 val descriptionValue = description
 val latestVersionValue = latestVersion
 val ownerIdValue = if (includes.contains("ownerId")) ownerId else null
 val isPublicValue = isPublic
 val isDeletedValue = isDeleted
 val contentHashValue = if (includes.contains("contentHash")) contentHash else null
 val createdAtValue = if (includes.contains("createdAt")) createdAt.toString() else null
 val updatedAtValue = updatedAt.toString()
 
 return ContractDAO(
     id = idValue,
     name = nameValue,
     description = descriptionValue,
     contentHash = contentHashValue,
     latestVersion = latestVersionValue,
     ownerId = ownerIdValue,
     isPublic = isPublicValue,
     isDeleted = isDeletedValue,
     createdAt = createdAtValue,
     updatedAt = updatedAtValue
 )
}

File: threadify-go/api/internal/dto/contract.go
package dto

import (
	"encoding/json"
	"time"
)

type Contract struct {
	ID            string    `json:"id"`
	Name          string    `json:"name"`
	CompanyID     string    `json:"companyId"`
	Description   string    `json:"description"`
	ContentHash   *string   `json:"contentHash"`
	LatestVersion int       `json:"latestVersion"`
	OwnerID       string    `json:"ownerId"`
	IsPublic      bool      `json:"isPublic"`
	IsDeleted     bool      `json:"isDeleted"`
	CreatedAt     time.Time `json:"createdAt"`
	UpdatedAt     time.Time `json:"updatedAt"`
}

type ContractVersion struct {
	ID                 string          `json:"id"`
	Version            int             `json:"version"`
	Content            string          `json:"content"`
	YAMLContent        string          `json:"yamlContent"`
	ContentHash        string          `json:"contentHash"`
	ContractID         string          `json:"contractId"`
	CreatedBy          string          `json:"createdBy"`
	Graph              json.RawMessage `json:"graph"`
	ExpectedDurationMs *int64          `json:"expectedDurationMs"`
	IsDeleted          bool            `json:"isDeleted"`
	CreatedAt          time.Time       `json:"createdAt"`
	UpdatedAt          time.Time       `json:"updatedAt"`
}

type ContractListResponse struct {
	Contracts []*Contract `json:"contracts"`
	Total     int         `json:"total"`
}

type ContractWithOwnershipResponse struct {
	Contract *Contract `json:"contract"`
	IsOwner  bool      `json:"isOwner"`
}

type ContractVersionsResponse struct {
	Versions []*ContractVersion `json:"versions"`
	Name     string             `json:"name"`
	Total    int                `json:"totalVersions"`
}
